lib/Doctrine/ORM/AbstractQuery.php

```diff
@@ -417,11 +417,19 @@ public function processParameterValue($value)
             return $value;
         }
 
-        if (is_object($value) && $this->_em->getMetadataFactory()->hasMetadataFor(ClassUtils::getClass($value))) {
-            $value = $this->_em->getUnitOfWork()->getSingleIdentifierValue($value);
+        if (is_object($value)) {
+            try {
+                $this->_em->getMetadataFactory()->getMetadataFor(ClassUtils::getClass($value));
```

Comment:
I think you can safely replace this call by the contents of the `if (!isset($e))` here. `UnitOfWork::getSingleIdentifierValue()` will call `ClassMetadataFactory::getMetadataFor()` right away (and throw if class is not a valid entity).